### PR TITLE
Added constrained planning support to OMPLPlanner

### PR DESCRIPTION
**Pull request by jeking04**
This PR adds the ability to do constrained planning with OMPL via the ```pr-constraint``` package developed by @cdellin.

### CODE REVIEW

File: src/prpy/planning/ompl.py [hunk 1]
```diff
@@ -245,3 +246,100 @@ def ShortcutPath(self, robot, path, timeout=1., **kwargs):
...
+        @param robot
+        @param tsrchains A list of TSRChain objects to respect during planning
+        @param planner_range the max stepsize for the planner before a projection
+         is applied
+        @param ompl_args ompl RRTConnect specific parameters
+        @return traj
+        """
+        if ompl_args is None:
+            ompl_args = {}
+        ompl_args['range'] = planner_range
```
Comment (mkoval) [inline src/prpy/planning/ompl.py]: Why do we manually set `range` only for constrained planners?

Also, I think the default value should be derived from DOF resolution. We have some logic to do this in the `RRTConnectPlanner` wrapper class.

File: src/prpy/planning/ompl.py [hunk 2]
```diff
@@ -245,3 +246,100 @@ def ShortcutPath(self, robot, path, timeout=1., **kwargs):
...
+                          manip = robot.GetActiveManipulatorIndex())
+
+            traj_tsr_chain = TSRChain(constrain=True, TSRs=[trajtsr])
+
+            # Compute a goal pose
+            goal_pose = manip.GetEndEffectorTransform()
+            goal_pose[:3,3] += direction*distance/numpy.linalg.norm(direction)
+
+            # Get all IKs for the goal pose and send them all to the planner
+            goal_configs = manip.FindIKSolutions(goal_pose, 0)
```
Comment (mkoval) [inline src/prpy/planning/ompl.py]: I thought `or_ompl` supported goal sets. Why do we manually find IK solutions here?

File: src/prpy/planning/ompl.py (hunk 1 above)
Comment (cdellin) [inline src/prpy/planning/ompl.py]: From what I remember (and I'll have to take a closer look to determine this more closely), the goal here was to emulate the way CBiRRT places vertices in its RRT after each projection.  Since the default `range` parameter is rather large, we currently overwrite it only for constrained planners.

**Comment by mkoval** [comment]
This is ready for review. @jeking04 can you take a look?

This depends on https://github.com/personalrobotics/or_ompl/pull/72.

File: src/prpy/planning/adapters.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class PlanToEndEffectorOffsetTSRAdapter(Planner):
+    epsilon = 1e-3
+
+
+    """ Plan PlanToEndEffectorOffset using PlanToTSR on a delegate planner.
+
+    @param delegate_planner planner that supports goal and constraint TSRs
+    """
```
Comment (gilwoolee) [inline src/prpy/planning/adapters.py]: This docstring needs to be under method declaration

File: src/prpy/planning/adapters.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    def __init__(self, delegate_planner):
+        super(PlanToEndEffectorOffsetTSRAdapter, self).__init__()
+
+        self.delegate_planner = delegate_planner
+        self.env = Environment()
+
+
+    """ Plan to a desired end-effector offset with fixed orientation.
```
Comment (gilwoolee) [inline src/prpy/planning/adapters.py]: This docstring needs to be under method declaration

File: src/prpy/planning/ompl.py [hunk 5]
```diff
@@ -81,7 +81,7 @@ class OMPLPlanner(BasePlanner):
     @param ompl_args dictionary of planner-specific arguments
```
Comment (gilwoolee) [inline src/prpy/planning/ompl.py]: Add docstring for `support_constraints`

File: src/prpy/planning/ompl.py [hunk 6]
```diff
@@ -175,12 +176,22 @@ def _Plan(self, robot, goal=None, tsrchains=None, timelimit=None,
...
+                        'OMPL does not support start TSRs.')
+
+                if chain.sample_goal:
+                    extraParams += '<{k:s}>{v:s}</{k:s}>'.format(
+                        k='tsr_chain', v=SerializeTSRChain(chain))
+
+                if chain.constrain:
+                    if self.supports_constraints:
+                        extraParams += '<{k:s}>{v:s}</{k:s}>'.format(
+                            k='tsr_chain_constraint', v=chain.to_json())
```
Comment (gilwoolee) [inline src/prpy/planning/ompl.py]: How come this uses `chain.to_json` and above one use `SerializeTSRChain`?

File: src/prpy/planning/cbirrt.py [hunk 7]
```diff
@@ -102,76 +107,30 @@ def PlanToEndEffectorPose(self, robot, goal_pose, **kw_args):
...
-                              [-epsilon,            epsilon]])
-
-            traj_tsr = prpy.tsr.tsr.TSR(
-                T0_w=H_world_w, Tw_e=H_w_ee, Bw=Bw,
-                manip=robot.GetActiveManipulatorIndex())
-            traj_tsr_chain = prpy.tsr.tsr.TSRChain(constrain=True,
-                                                   TSRs=[traj_tsr])
+        chains = PlanToEndEffectorOffsetTSRAdapter.CreateTSRChains(
+                    robot, direction, distance)
+        return self.PlanToTSR(robot, tsr_chains=chains, **kw_args)
```
Comment (gilwoolee) [inline src/prpy/planning/cbirrt.py]: The previous version had `kw_args.setdefault('psample', 0.1)`. Is this no longer necessary?

Comment (mkoval) [inline src/prpy/planning/cbirrt.py]: This function now calls `PlanToTSR`, which has the same `setdefault('psample', 0.1)` line in it.

File: src/prpy/planning/ompl.py (hunk 6 above)
Comment (mkoval) [inline src/prpy/planning/ompl.py]: I added a comment to address this.

File: src/prpy/planning/ompl.py [hunk 8]
```diff
@@ -175,12 +180,26 @@ def _Plan(self, robot, goal=None, tsrchains=None, timelimit=None,
...
+                    extraParams += '<{k:s}>{v:s}</{k:s}>'.format(
+                        k='tsr_chain', v=SerializeTSRChain(chain))
+
+                # Constraint TSRs are parsed by pr_constraint_or_ompl, which
+                # uses a JSON format. This differs from the space delimited
+                # format used by CBiRRT.
+                if chain.constrain:
+                    if self.supports_constraints:
+                        extraParams += '<{k:s}>{v:s}</{k:s}>'.format(
+                            k='tsr_chain_constraint', v=chain.to_json())
```
Comment (gilwoolee) [inline src/prpy/planning/ompl.py]: Why does this use `chain.to_json` and two lines above it's `SerializeTSRChain`?

Comment (mkoval) [inline src/prpy/planning/ompl.py]: It's just how it is. `or_ompl` (which uses the CBiRRT format) deserializes goal TSRs and `pr_constraint_or_ompl` (which uses JSON) deserializes constraint TSRs.

Comment (psigen) [inline src/prpy/planning/ompl.py]: Can we port or_ompl to use the json format as well?

Comment (mkoval) [inline src/prpy/planning/ompl.py]: I am fine with adding the JSON format as an option. That change belongs in a separate pull request.

If we do so, I would like to keep `<tsr_chain>` using the CBiRRT format for backwards compatibility.

Comment (mkoval) [inline src/prpy/planning/ompl.py]: I created an issue to track this here: https://github.com/personalrobotics/or_ompl/issues/73.

**Comment by mkoval** [comment]
@gilwoolee signed off on this so I am merging it.

File: src/prpy/planning/ompl.py (hunk 8 above)
Comment (psigen) [inline src/prpy/planning/ompl.py]: Sounds good. Thanks.
